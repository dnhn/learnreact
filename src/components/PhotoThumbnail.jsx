import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { openPhoto } from '../redux/actions';

import './PhotoThumbnail.scss';

class PhotoThumbnail extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      show: false,
    };
  }

  componentDidMount() {
    setTimeout(() => {
      this.setState({
        show: true,
      });
    }, this.props.order * 150);
  }

  render() {
    const {
      data: photoData,
      openPhoto,
    } = this.props;
    const {
      links,
      urls,
      description,
    } = photoData;
    const { show } = this.state;

    return (
      <a
        href={links.html}
        target="_blank"
        rel="noopener noreferrer"
        className={`photo ${show ? 'show' : ''}`}
        onClick={e => { e.preventDefault(); openPhoto(photoData) }}
        style={{backgroundImage: `url(${urls.small})`}}>
        {description ? <span className="photo__desc">{description}</span> : ''}
      </a>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  openPhoto: selectedPhoto => dispatch(openPhoto(selectedPhoto)),
});

export default connect(null, mapDispatchToProps)(PhotoThumbnail);
