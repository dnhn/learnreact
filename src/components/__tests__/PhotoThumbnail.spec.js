import React from 'react';
import { fireEvent, render } from "@testing-library/react";

import { PhotoThumbnail } from '../PhotoThumbnail';

let mockStore = {};

beforeEach(() => {
  mockStore = {
    data: {
      links: { html: '#link' },
      urls: { small: '#background' },
      description: 'description',
    },
    openPhoto: jest.fn(),
  };
});

describe('<PhotoThumbnail />', () => {
  it('renders without crashing', () => {
    const { asFragment } = render(<PhotoThumbnail {...mockStore} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('has blank description', () => {
    mockStore.data.description = '';
    const { container } = render(<PhotoThumbnail {...mockStore} />);
    expect(container.getElementsByClassName('photo__desc').length)
      .toBe(0);
  });

  it('is clicked', () => {
    const { container } = render(<PhotoThumbnail {...mockStore} />);
    fireEvent.click(container.getElementsByClassName('photo')[0]);
    expect(mockStore.openPhoto).toHaveBeenCalled();
  });
});
