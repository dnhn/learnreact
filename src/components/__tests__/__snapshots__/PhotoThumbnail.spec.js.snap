// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoThumbnail /> renders without crashing 1`] = `
<DocumentFragment>
  <a
    class="photo "
    href="#link"
    rel="noopener noreferrer"
    style="background-image: url(#background);"
    target="_blank"
  >
    <span
      class="photo__desc"
    >
      description
    </span>
  </a>
</DocumentFragment>
`;
